### PR TITLE
ci: migrate release process to nix (PROOF-537)

### PR DESCRIPTION
**Pull request by rnburn**
# Rationale for this change

Migrate the release process to use nix.

Note: After this is merged, we can remove the docker files and simplify the linking in bazel. I plan to follow up with a PR to clean up the environment and update the READMEs.

# What changes are included in this PR?

* Changed the binary to link against libc++ instead of libstdc++ to work around to https://github.com/llvm/llvm-project/issues/69288
* Sets up an older version of gcc and glibc to link against so that we can make a portable binary. The previous approach used symbol remapping which is no longer necessary.
* Migrates rust tests and semantic release to use nix.
* Removes the docker portion flake.nix as we ended up not using that
* Adds -no-pie as a workaround to https://github.com/google/sanitizers/issues/856 which comes up when we use ASAN with an older version of glibc
* Sets up builds with different options to use different bazel cache directories so that CI jobs will run faster

# Are these changes tested?

The usual tests are run in CI with the new environment. I manually inspected the binary libblitar.so with `objdump -T` and `ldd` to check that it doesn't link to anything that produces a portability problem.

The changes to semantic release haven't been tested. I'm hoping that they will work without additional tweaks, but if they pose a problem I'll either follow on with a fix or rollback the PR to rework the semantic release job.

### CODE REVIEW

File: nix/shell.nix [hunk 1]
```diff
@@ -7,8 +7,15 @@ in
...
-    pkgs.python3
+    bazel-buildtools
+    python3
+    rust-bin.nightly."2023-12-01".default
+    # cargo
+    rust-bindgen
+    # rustfmt
+    patchelf
+    nodejs
+    # custom
```
Comment (jacobtrombetta) [inline nix/shell.nix:18]: Should `cargo`, `rustfmt`, or `custom` be comments?

File: .github/workflows/test-check-lint.yml [hunk 2]
```diff
@@ -19,23 +19,23 @@ jobs:
...
     steps:
       - name: Checkout Code
         uses: actions/checkout@v3
-
+  
       - name: Run cpp tests
         run: >
-          nix develop --command bazel test //...
-
+          TEST_TMPDIR=$HOME/.bazel_test nix develop --command bazel test //...
```
Comment (jacobtrombetta) [inline .github/workflows/test-check-lint.yml:37]: Will this line create `$HOME/.bazel_test` if the folder doesn't exist? I see in `release.yml` the line `run: mkdir -p $HOME/.bazel_test_release` before `TEST_TMPDIR=$HOME/.bazel_test_release` is used. Why is the `mkdir` called needed in `release.yml` and not in `test-check-lint.yml`?

The same question would apply to `bazel_test_opt` and `bazel_test_asan`.

File: nix/shell.nix (hunk 1 above)
Comment (rnburn) [reply nix/shell.nix:18]: Sorry. Let me remove "custom" was meant to be a descriptive comments; I'll delete the others.

File: .github/workflows/test-check-lint.yml (hunk 2 above)
Comment (rnburn) [reply .github/workflows/test-check-lint.yml:37]: Yes. It creats that. I'll remove the mkdir command -- it was something I was trying while putting things together.

**Review by jacobtrombetta — APPROVED** [review]

**Pull request merged**